work/work.go

```diff
@@ -0,0 +1,191 @@
+		return Report{}, errs.New("run timeout too short")
+	}
+
+	tracer := newTracer(logger, w.AgentFlushTimeout, w.SecretToken, w.ServerUrl, w.MaxSpansPerTransaction)
+
+	w.Add(errors(throttle(w.ErrorFrequency), tracer, w.ErrorLimit, w.MinFramesPerError, w.MaxFramesPerError))
+	w.Add(transactions(throttle(w.TransactionFrequency), tracer, w.TransactionLimit, w.MinSpansPerTransaction, w.MaxSpansPerTransaction))
+	w.Add(timeout(w.RunTimeout))
+
+	logger.Debugf("start")
+	defer logger.Debugf("finish")
+
+	report := Report{Start: time.Now()}
+	err := w.Run()
+	tracer.flush()
+	report.End = time.Now()
+
+	stats := tracer.Stats()
+	report.TransactionsSent = stats.TransactionsSent
+	report.SpansSent = stats.SpansSent
+	report.ErrorsSent = stats.ErrorsSent
+	report.RequestErrors = stats.Errors.SendStream
+
+	return report, err
+}
+
+// throttle converts a time ticker to a channel of things
+func throttle(d time.Duration) chan interface{} {
+	throttle := make(chan interface{})
+	go func() {
+		for range time.NewTicker(d).C {
+			throttle <- struct{}{}
+		}
+	}()
+	return throttle
+}
+
+type generator func(<-chan struct{}) error
+
+func transactions(throttle <-chan interface{}, tracer *Tracer, limit, spanMin, spanMax int) generator {
+	if limit <= 0 {
+		return noop()
+	}
+
+	generateSpan := func(ctx context.Context) {
+		span, ctx := apm.StartSpan(ctx, "I'm a span", "gen.era.ted")
+		span.End()
+	}
+	return func(done <-chan struct{}) error {
+		sent := 0
+		for sent < limit {
+			select {
+			case <-done:
+				return nil
+			case <-throttle:
+			}
+
+			tx := tracer.StartTransaction("generated", "gen")
+			ctx := apm.ContextWithTransaction(context.Background(), tx)
+			var wg sync.WaitGroup
+			spanCount := rand.Intn(spanMax-spanMin+1) + spanMin
+			for i := 0; i < spanCount; i++ {
+				wg.Add(1)
+				go func() {
+					generateSpan(ctx)
+					wg.Done()
+				}()
+			}
+			wg.Wait()
+			tx.Context.SetTag("spans", strconv.Itoa(spanCount))
+			tx.End()
+			sent = int(tracer.Stats().TransactionsSent)
+		}
+		return nil
+	}
+
+}
+
+type generatedErr struct {
+	frames int
+}
+
+func (e *generatedErr) Error() string {
+	plural := "s"
+	if e.frames == 1 {
+		plural = ""
+	}
+	return fmt.Sprintf("Generated error with %d stacktrace frame%s", e.frames, plural)
+}
+
+func (e *generatedErr) StackTrace() []stacktrace.Frame {
+	st := make([]stacktrace.Frame, e.frames)
+	for i := 0; i < e.frames; i++ {
+		st[i] = stacktrace.Frame{
+			File:     "fake.go",
+			Function: "oops",
+			Line:     i + 100,
+		}
+	}
+	return st
+}
+
+func errors(throttle <-chan interface{}, tracer *Tracer, limit, framesMin, framesMax int) generator {
+	if limit <= 0 {
+		return noop()
+	}
+	return func(done <-chan struct{}) error {
+		sent := 0
+		for sent < limit {
+			select {
+			case <-done:
+				return nil
+			case <-throttle:
+			}
+			tracer.NewError(&generatedErr{frames: rand.Intn(framesMax-framesMin+1) + framesMin}).Send()
+			sent = int(tracer.Stats().ErrorsSent)
```

Comment:
is this accurate with the number of errors generated or sent?  Any worries about contention on this lock preventing load generation from hitting max throughput?

Reply:
not that i observed, but didn't test thoroughly. i can put back the counter